ARF Video Source

aaron-bray opened the pull request:
Here is a basic working implementation
With arf videos recorded by our team
I have the specification and may go back through and fill out more metadata and support.
I could also go through testing more arf files to ensure the various encodings are properly implemented..

The MAIN question I have is if the video source should hold onto any of the images
Currently, it just holds a pointer to the file, and when a frame is requested, it creates a kv::image and returns it
Which I think is the correct KWIVER behavior

Note that the simple image container returned also able to be successfully transformed into ocv using the vital_to_ocv function

Let me know what you think!

mwoehlke-kitware commented:
> The MAIN question I have is if the video source should hold onto any of the images

Not more than the most recently requested frame; otherwise you're either adding extra complexity for questionable benefit (to do some sort of bounded cache), or risking unbounded memory growth.

Caching the last frame, however, may be useful in case of stupid users that call `frame_image` multiple times without any intervening seeks.

mwoehlke-kitware commented:
Please consider running uncrustify on this to improve style consistency. Also, please at least fix the length of your separator lines. Avoiding over-length lines would be nice also.

mleotta commented on arrows/core/video_input_arf.cxx:
```diff
@@ -0,0 +1,558 @@
+    d->px.num_bytes = d->bpp;
+    d->px.type = kv::image_pixel_traits::UNSIGNED;
+    break;
+  case 3://
```
same as case 5 above?

mleotta commented on arrows/core/video_input_arf.cxx:
```diff
@@ -0,0 +1,558 @@
+  {
+  case 0:
+    d->bpp = 1;//uint8
+    d->px.num_bytes = d->bpp;
```
it looks like setting `num_bytes` is done the same way in all cases, so you could pull this out of the switch.

aaron-bray commented:
Good point on the current frame cache

Will also do the uncrustification/styling

mwoehlke-kitware commented on arrows/core/video_input_arf.cxx:
```diff
@@ -0,0 +1,558 @@
+  unsigned short byteswap(unsigned short);
+  unsigned long byteswap(unsigned long);
+  unsigned int byteswap(unsigned int);
+  unsigned __int64 byteswap(unsigned __int64);
```
This is generating a compile error:
```
kwiver/arrows/core/video_input_arf.cxx:54:12: error: expected ‘;’ at end of member declaration
```

It looks like the compiler it interpreting this (unsurprisingly) as a member named `__int64`.

Please either use `long long`, or, since you already use `uint32_t` elsewhere, consider `uint64_t`, and `uintN_t` also for the others.

mwoehlke-kitware commented on arrows/core/video_input_arf.cxx:
```diff
@@ -0,0 +1,558 @@
+#ifdef WIN32
+  return _byteswap_ulong(value);
+#else
+  return __builtin_bswap32(value);
```
This is just plain wrong on 64-bit platforms. In fact, why do you even need this overload?

It looks like the only one you *actually* need is the `uint32_t` overload. Also, while we probably don't especially care about big-endien platforms, I suspect you still actually want to use `htonl`/`ntohl` instead, which have the benefit of being more portable.

mwoehlke-kitware commented on arrows/core/video_input_arf.cxx:
```diff
@@ -0,0 +1,558 @@
+namespace arrows {
+namespace core {
+
+class video_input_arf::priv
```
```suggestion
// ----------------------------------------------------------------------------
class video_input_arf::priv
```

mleotta commented on arrows/core/video_input_arf.h:
```diff
@@ -0,0 +1,75 @@
+// (Automatic Target Recognition Working Group Raster Format)
+// ----------------------------------------------------------------
+/**
+ * This class implements a video input algorithm that returns only metadata.
```
This comment does not look correct

mwoehlke-kitware commented on arrows/core/video_input_arf.cxx:
```diff
@@ -0,0 +1,558 @@
+  bool arf_comment = (flags & (2 ^ 2)) != 0;
+  bool arf_multiband = (flags & (2 ^ 3)) != 0;
+  bool arf_framedata = (flags & (2 ^ 4)) != 0;
+  bool arf_extra = (flags & (2 ^ 5 - 1)) != 0;
```
```
kwiver/arrows/core/video_input_arf.cxx:170:36: warning: suggest parentheses around arithmetic in operand of ‘^’ [-Wparentheses]
  170 |   bool arf_extra = (flags & (2 ^ 5 - 1)) != 0;
      |                                  ~~^~~
```

mleotta commented on arrows/core/video_input_arf.cxx:
```diff
@@ -0,0 +1,558 @@
+video_input_arf
+::frame_metadata()
+{
+  kv::metadata_vector vect;
```
If there is no metadata, I think we can return an empty vector.  If there is code elsewhere that crashes if this vector is empty we should probably fix that.

mwoehlke-kitware commented on arrows/core/video_input_arf.cxx:
```diff
@@ -0,0 +1,558 @@
+video_input_arf
+::open( std::string filename)
+{
+  typedef kwiversys::SystemTools ST;
+
```
```suggestion
```

(...and, anyway, please don't use `typedef`. Ever. :slightly_smiling_face:)

mleotta commented on arrows/core/video_input_arf.cxx:
```diff
@@ -0,0 +1,558 @@
+video_input_arf
+::metadata_map()
+{
+  return std::make_shared< kv::simple_metadata_map >(d->metadata_map);
```
Likewise, I think you can return `nullptr` here if you don't actually have metadata.

mwoehlke-kitware commented on arrows/core/video_input_arf.cxx:
```diff
@@ -0,0 +1,558 @@
+  case 10:
+    d->bpp = 3;// uint8
+    d->px.num_bytes = d->bpp;
+    d->px.type = kv::image_pixel_traits::UNSIGNED;
```
My compiler thinks you forgot a `break;`?

mwoehlke-kitware reviewed (changes requested):
This is just a quick first pass, as right now the code does not even compile...

mleotta reviewed (changes requested):
Seems like a reasonable approach overall.  A unit test or two would be helpful if we have a small public ARF test file or can make one.  Also see inline comments.

mwoehlke-kitware commented on arrows/core/video_input_arf.cxx:
```diff
@@ -0,0 +1,558 @@
+video_input_arf
+::frame_metadata()
+{
+  kv::metadata_vector vect;
+  if ( d->metadata )
+  {
+    vect.push_back( d->metadata );
+  }
+
+  return vect;
```
Per @mleotta's comment:
```suggestion
  return {};
```

mwoehlke-kitware commented on arrows/core/video_input_arf.cxx:
```diff
@@ -0,0 +1,558 @@
+// ----------------------------------------------------------------------------
+void
+video_input_arf::priv
+::byteswap_img_bytes(unsigned char* img_bytes)
```
I'm not sure if it's worth rewriting this (maybe later?), but I can't help but wonder if rewriting this to iterate over a `rows*cols` integer array using `ntohX` wouldn't be faster. (It generates radically different, and shorter, assembly, at least...)

mwoehlke-kitware commented on arrows/core/video_input_arf.cxx:
```diff
@@ -48,10 +48,10 @@ class video_input_arf::priv
-  unsigned long byteswap(unsigned long);
-  unsigned int byteswap(unsigned int);
-  unsigned __int64 byteswap(unsigned __int64);
+  uint8_t byteswap(uint8_t);
```
This overload is meaningless. All but one of them continue to be unused.

mleotta commented:
@aaron-bray I'm wondering about the status of this PR.  There are a lot of unaddressed comments.  Do you still plan to address them?

aaron-bray commented:
I do, I will get to this next week (hopefully)

aaron-bray commented:
I believe I have addressed most issues
Still works fine on our project
As for a unit test, these are uncompressed video's and very large
I do not have any video smaller that 10MB to include in the repo to use as test input

aaron-bray commented:
@mleotta 
Is this looking square to you?

aaron-bray commented:
@mleotta @mwoehlke-kitware 

Ping

aaron-bray commented:
@mwoehlke-kitware @mleotta You guys want this in the 1.6 release?

mleotta commented:
Good question.  I think it's more a question of whether you want it in the 1.6 release.  Either way we should get it landed.  Unfortunately, I'm not going to have time to do another detailed review, I've got too much else going on.

aaron-bray commented:
allright
Functionally, this code is working well with arf videos from our camera
I also implemenet all the suggested functionality

Whats left is pretty much styling.
I don't have the time to hunt down a bunch of other arf files that may vary in encoding to test those aspects of the code
I think this is a good start, and it would be pretty easy and quick to fix any bugs if another encoding happed to not work...
